### ISSUE TITLE
How to scheduler.step() after every batch

### ISSUE THREAD

**Issue by benihime91** [PyTorchLightning/pytorch-lightning#3051]
I want to make use of schedulers like `torch.optim.lr_scheduler.CyclicLR` & `torch.optim.lr_scheduler.OneCycleLR` 
https://pytorch.org/docs/stable/optim.html
These schedulers require `scheduler.step()` to be callled after each `batch`. 
How to achieve this is `PyTorchLightning`?

**Comment by rohitgr7** [comment]
https://pytorch-lightning.readthedocs.io/en/latest/lightning-module.html#configure-optimizers

just set the `interval`:
```python
scheduler = {
    'scheduler': scheduler,
    'interval': 'step', # or 'epoch'
    'frequency: 1
}
```

**Comment by benihime91** [comment]
@rohitgr7 Thanks

**Comment by jmerkow** [comment]
> https://pytorch-lightning.readthedocs.io/en/latest/lightning-module.html#configure-optimizers
> 
> just set the `interval`:
> 
> ```python
> scheduler = {
>     'scheduler': scheduler,
>     'interval': 'step', # or 'epoch'
>     'frequency: 1
> }
> ```



Is there an updated link for this?

**Comment by rohitgr7** [comment]
@jmerkow just search for `configure_optimizers` in the search bar of the docs.
https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#configure-optimizers

**Comment by msverma101** [comment]
i want to use the the lr scheduler for gan and i am trying to figure out where to give the lr_scheduler_dis.step()
if i add it in training_step after loss is updated i am getting an warning which i want to fix
Warning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate


    def _gen_step(self, real_images, conditioned_images, scalar_gen, scalar_dis, aux_labels):
        # Pix2Pix has adversarial and a reconstruction loss
        # First calculate the adversarial loss
        # gen_opt = self.optimizers()

        fake_images = self.gen(conditioned_images, scalar_gen)
        dis_logits, aux_output = self.dis(fake_images, conditioned_images, scalar_dis)

        adversarial_loss = self.adversarial_criterion(dis_logits, torch.ones_like(dis_logits))
        criterion_loss = self.criterion(aux_output, aux_labels)#labels need to add
        error_fake = adversarial_loss + criterion_loss #adding real loss and criterion loss from labels
        error_fake.backward(retain_graph=True)#here not sure where to place this
        
        self.log("6_error_fake_gen", error_fake)
        self.log("2_Generator_adver_Loss", adversarial_loss)

        # calculate reconstruction loss
        recon_loss = self.recon_criterion(fake_images, real_images)
        lambda_recon = self.lambda_recon
        self.log("3_Generator_recon_Loss", recon_loss)

        gen_loss = adversarial_loss + lambda_recon * recon_loss

        # gen_opt.zero_grad()
        # self.manual_backward(gen_loss)
        # gen_opt.step()

        return gen_loss

    

    def _dis_step(self, real_images, conditioned_images, scalar_gen, scalar_dis, aux_labels):
        # dis_opt = self.optimizers()#turinign of auto opt need to specify opt manually
        fake_images = self.gen(conditioned_images, scalar_gen)
        fake_logits, _ = self.dis(fake_images, conditioned_images, scalar_dis)

        real_logits, aux_output = self.dis(real_images, conditioned_images, scalar_dis)

        fake_loss = self.adversarial_criterion(fake_logits, torch.zeros_like(fake_logits))
        real_loss = self.adversarial_criterion(real_logits, torch.ones_like(real_logits))

        criterion_loss = self.criterion(aux_output, aux_labels)#labels need to add
        error_real = real_loss + criterion_loss #adding real loss and criterion loss from labels
        error_real.backward(retain_graph=True)#here not sure where to place this
        self.log("7_error_real_dis", error_real)

        accuracy = self.compute_acc(aux_output, aux_labels)

        self.log("4_Discriminator_fake_Loss", fake_loss)
        self.log("5_Discriminator_real_Loss", real_loss)
        self.log("7_accruracy", accuracy)

        dis_loss = (real_loss + fake_loss) / 2

        # dis_opt.zero_grad()
        # self.manual_backward(dis_loss)
        # dis_opt.step()
        return dis_loss #, error_real

    def configure_optimizers(self):
        dis_opt = torch.optim.Adam(self.dis.parameters(), lr=self.learning_rate) 
        gen_opt = torch.optim.Adam(self.gen.parameters(), lr=self.learning_rate)
        #learning rate scheduler
        #https://medium.com/mlearning-ai/make-powerful-deep-learning-models-quickly-using-pytorch-lightning-29f040158ef3
        self.lr_scheduler_dis = torch.optim.lr_scheduler.StepLR(dis_opt, step_size  = 10 , gamma = 0.5)
        self.lr_scheduler_gan = torch.optim.lr_scheduler.StepLR(gen_opt, step_size  = 10 , gamma = 0.5)
        
        return dis_opt, gen_opt
